js/packages/openinference-instrumentation-bedrock-agent-runtime/src/callbackHandler.ts

```diff
@@ -67,3 +70,61 @@ export class CallbackHandler {
     this.span.end();
   }
 }
+
+/**
+ * RagCallbackHandler is responsible for handling streaming RAG (Retrieve and Generate)
+ * responses and citations during Bedrock agent runtime instrumentation.
+ *
+ * This handler accumulates output and citation data as the RAG operation progresses,
+ * and sets the appropriate OpenTelemetry span attributes upon completion or error.
+ *
+ * - handleCitation: Collects citation objects as they are received.
+ * - onComplete: Sets output and citation attributes on the span and ends it.
+ * - onError: Records exceptions and sets error status on the span.
+ */
+export class RagCallbackHandler {
+  private output: string;
+  private citation: Citation[];
+  private readonly span: Span;
+
+  /**
+   * Callback handler for processing agent responses and traces.
+   * @param span {Span} - The OpenTelemetry span to associate with the response.
+   */
+  constructor(span: Span) {
+    this.span = span;
+    this.output = "";
+    this.citation = [];
+  }
+
+  handleOutput(output: string) {
+    this.output = this.output + output;
+  }
+  handleCitation(citation: Record<string, unknown>) {
+    if (citation && Object.keys(citation).length > 0) {
+      this.citation.push(citation);
+    }
+  }
+  onComplete(): void {
+    this.span.setAttributes({
+      ...getOutputAttributes(this.output),
+      ...extractRetrievedReferencesAttributes(this.citation),
+    });
+    this.span.setStatus({ code: SpanStatusCode.OK });
+    this.span.end();
+  }
+  onError(error: unknown): void {
+    if (error instanceof Error) {
+      this.span.recordException(error);
+    }
+    const message =
+      error &&
+      typeof error === "object" &&
+      "message" in error &&
+      typeof error.message === "string"
+        ? error.message
+        : String(error);
```

Comment:
safely stringify here instead of String()


```suggestion
        : safelyJSONStringify(error);
```